src/gadgets/bigint/add.rs

```diff
@@ -0,0 +1,330 @@
+    BigIntWires { bits }
+}
+
+pub fn sub_generic_without_borrow(
+    circuit: &mut Circuit,
+    a: &BigIntWires,
+    b: &BigIntWires,
+) -> BigIntWires {
+    let BigIntWires { mut bits } = sub_generic(circuit, a, b);
+    bits.pop();
+    BigIntWires { bits }
+}
+
+pub fn double(circuit: &mut Circuit, a: &BigIntWires) -> BigIntWires {
+    let zero_wire = circuit.issue_wire();
+    circuit.add_gate(Gate::nimp(a.bits[0], a.bits[0], zero_wire));
+
+    BigIntWires {
+        bits: iter::once(zero_wire)
+            .chain(a.bits.iter().copied())
+            .collect(),
+    }
+}
+
+//    pub fn double_without_overflow(a: Wires) -> Circuit {
+//        assert_eq!(a.len(), N_BITS);
+//        let mut circuit = Circuit::empty();
+//        let not_a = new_wirex();
+//        let zero_wire = new_wirex();
+//        circuit.add(Gate::not(a[0].clone(), not_a.clone()));
+//        circuit.add(Gate::and(a[0].clone(), not_a.clone(), zero_wire.clone()));
+//        circuit.add_wire(zero_wire);
+//        circuit.add_wires(a[0..N_BITS - 1].to_vec());
+//        circuit
+//    }
+pub fn double_without_overflow(circuit: &mut Circuit, a: &BigIntWires) -> BigIntWires {
+    let zero_wire = circuit.issue_wire();
+    circuit.add_gate(Gate::nimp(a.bits[0], a.bits[0], zero_wire));
+
+    BigIntWires {
+        bits: iter::once(zero_wire)
+            .chain(a.bits.iter().take(a.bits.len() - 1).copied())
+            .collect(),
+    }
+}
+
+pub fn half(circuit: &mut Circuit, a: &BigIntWires) -> BigIntWires {
+    let zero_wire = circuit.issue_wire();
+    circuit.add_gate(Gate::nimp(a.bits[0], a.bits[0], zero_wire));
+
+    BigIntWires {
+        bits: a
+            .bits
+            .iter()
+            .skip(1)
+            .copied()
+            .chain(iter::once(zero_wire))
+            .collect(),
+    }
+}
+
+pub fn odd_part(circuit: &mut Circuit, a: &BigIntWires) -> (BigIntWires, BigIntWires) {
+    let mut select_bn = BigIntWires::new(circuit, a.len() - 1, false, false);
+    select_bn.insert(0, a.get(0).unwrap());
+
+    for i in 1..a.len() {
+        circuit.add_gate(Gate::or(
+            select_bn.get(i - 1).unwrap(),
+            a.get(i).unwrap(),
+            select_bn.get(i).unwrap(),
+        ));
+    }
+
+    let mut k = BigIntWires::new(circuit, a.len() - 1, false, false);
+    k.insert(0, a.get(0).unwrap());
+
+    for i in 1..a.len() {
+        circuit.add_gate(Gate::and_variant(
+            select_bn.get(i - 1).unwrap(),
+            a.get(i).unwrap(),
+            k.get(i).unwrap(),
+            [true, false, false],
+        ));
+    }
+
+    let mut odd_acc = a.clone(); // needs `Clone` on BigIntWires
+
+    for i in 1..a.len() {
+        let half_res = half(circuit, &odd_acc);
+        odd_acc = select(circuit, &odd_acc, &half_res, select_bn.get(i).unwrap());
+    }
+
+    (odd_acc, k)
+}
+
+#[cfg(test)]
+mod tests {
+    use rand::SeedableRng;
+    use test_log::test;
+
+    use super::*;
+
+    fn trng() -> rand::rngs::StdRng {
```

Comment:
we can create a util function for rng, like [create_rng](https://github.com/GOATNetwork/bitvm2-gc/blob/25b1062e582c78be6ad64140954f41c6bfc63afa/garbled-snark-verifier/src/circuits/bn254/utils.rs#L71C1-L91C2), and unify all the rng generator in other files.

Reply:
Done!